Improve SignalR Nginx settings (#19983)
dotnet/AspNetCore.Docs: aspnetcore/signalr/scale.md

## Linux with Nginx

The following contains the minimum required settings to enable WebSockets, ServerSentEvents, and LongPolling for SignalR:

```nginx
http {
  map $http_upgrade $connection_upgrade {
    default Upgrade;
    '' close;

>   }

Reviewer: What's going on here?

Author: Like I was saying on the issue, "Connection: close" is applied by default in Nginx, and "Connection: keep-alive" should not be forwarded by proxies as it is a hop-by-hop header. We can have another discussion about that behavior, but I think for this PR this is a big improvement over what we had before.

Reviewer: Now I remember. I have the same question [I had on the issue](https://github.com/dotnet/AspNetCore.Docs/issues/10827#issuecomment-631177400).

> Does this mean that every request to /hubroute will always contain either a "Connection: upgrade" or a "Connection: close" header? We wouldn't want to disable keep-alive for long polling.

Sure, "Connection" is a hop-by-hop header, but I want to make sure the connection between nginx and the signalr server isn't closed after every request.

Author: Is it safe to make all connections keep alive?

Reviewer: I don't see why not. Most modern HTTP proxies and clients do. It can be problem with stateful NTLM connections, but we're not telling people how to configure NTLM.

@Tratcher

Reviewer 2: Yes, enable keep-alive by default between nginx and kestrel.

  server {
    listen 80;
    server_name example.com *.example.com;

>     root /path/to/wwwroot;

Reviewer: Why bother with this in the signalr doc? What if wwwroot is on another machine? I expect static files will be proxied in many cases.

    # Configure the SignalR Endpoint
    location /hubroute {
      # App server url
      proxy_pass http://localhost:5000;

      # Configuration for WebSockets
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;

>       proxy_cache_bypass $http_upgrade;

Reviewer: If we worried about the proxy caching requests, shouldn't we be worried about every request, not just WebSocket requests?

```suggestion
      proxy_cache off;
```

      # Configuration for ServerSentEvents
      proxy_buffering off;

>       # Configuration for LongPolling

Reviewer: It looks like this is needed for WebSockets if the KeepAliveInterval is large enough http://nginx.org/en/docs/http/websocket.html

      proxy_read_timeout 100s;

      proxy_set_header Host $host;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
    }
  }
}
```

> When multiple backend servers are used sticky sessions must be added so SignalR connections do not switch servers when connecting.

Reviewer 3: I think you meant to combine sentences into a paragraph ...

```suggestion
When multiple backend servers are used, sticky sessions must be added to prevent SignalR connections from switching servers when connecting. There are multiple ways to add sticky sessions in Nginx. Two approaches are shown below depending on what you have available.

```

> There are multiple ways to add sticky sessions in Nginx, we show two below depending on what you have available to you.

Reviewer 3: ```suggestion
```

> The following is added in addition to the previous configuration:

Reviewer 3: Recommend moving as much of the unlocalized language into the text as possible.

```suggestion
The following is added in addition to the previous configuration. `backend` in the following example is the name of the group of servers:
```

```nginx
http {

>   # 'backend' is the name of our group of servers

Reviewer 3: ```suggestion
```

  upstream backend {
    # App server 1
    server http://localhost:5000;
    # App server 2
    server http://localhost:5002;

    # for Nginx Plus
    sticky cookie srv_id expires=max domain=.example.com path=/ httponly;

    # for Nginx Open Source
    ip_hash;
  }
}
```

> Finally, change `proxy_pass http://localhost:5000` in the `server {` section to `proxy_pass http://backend`.

Reviewer 3: ```suggestion
Finally, change `proxy_pass http://localhost:5000` in the `server` section to `proxy_pass http://backend`.
```

> For more information on WebSockets over Nginx, see [NGINX as a WebSocket Proxy](https://www.nginx.com/blog/websocket-nginx/).

Reviewer 3: Turns out for this link, they don't redirect, so we can go slash-less on it ...

```suggestion
For more information on WebSockets over Nginx, see [NGINX as a WebSocket Proxy](https://www.nginx.com/blog/websocket-nginx).
```

## Linux with Nginx

The following contains the minimum required settings to enable WebSockets, ServerSentEvents, and LongPolling for SignalR:

```nginx
http {
  map $http_upgrade $connection_upgrade {
    default Upgrade;

>     '' close;

Author: ```suggestion
    '' keep-alive;
```

  }

  server {
    listen 80;
    server_name example.com *.example.com;

    # Configure the SignalR Endpoint
    location /hubroute {
      # App server url
      proxy_pass http://localhost:5000;

      # Configuration for WebSockets
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
      proxy_cache off;

      # Configuration for ServerSentEvents
      proxy_buffering off;

      # Configuration for LongPolling
      proxy_read_timeout 100s;

      proxy_set_header Host $host;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
    }
  }
}
```

When multiple backend servers are used, sticky sessions must be added to prevent SignalR connections from switching servers when connecting. There are multiple ways to add sticky sessions in Nginx. Two approaches are shown below depending on what you have available.

> The following is added in addition to the previous configuration. `backend` in the following example is the name of the group of servers:

Reviewer 3: ```suggestion
The following is added in addition to the previous configuration. In the following examples, `backend` is the name of the group of servers.
```

*"Ooops!*" on this one. The conditional is better up-front.

> With Nginx Open Source, use `ip_hash` to route connections to a server based on the clients IP address.

Reviewer 3: ```suggestion
With [Nginx Open Source](https://nginx.org/en/), use `ip_hash` to route connections to a server based on the client's IP address:
```

Reviewer 3: btw - the "en" localization is required for their doc to load. **_Shame on them!_** :smile: lol

```nginx
http {
  upstream backend {
    # App server 1
    server http://localhost:5000;
    # App server 2
    server http://localhost:5002;

    ip_hash;
  }
}
```

> With Nginx Plus, use `sticky` to add a cookie to requests and pin them to a server.

Reviewer 3: ```suggestion
With [Nginx Plus](https://www.nginx.com/products/nginx), use `sticky` to add a cookie to requests and pin the user's requests to a server:
```

## Linux with Nginx

The following contains the minimum required settings to enable WebSockets, ServerSentEvents, and LongPolling for SignalR:

```nginx
http {
  map $http_connection $connection_upgrade {

>   "~*Upgrade" Upgrade;

Reviewer: Out of curiosity, does the following work?

```suggestion
  "~*Upgrade" $http_connection;
```

Author: It does

>     '' keep-alive;

Reviewer: What's the reason for basing the Connection header on the incoming Upgrade header? Maybe the following which bases it on the incoming Connection header would make more sense. It seems more straightforward not to look at a different header.

```
map $http_connection $connection_upgrade {
  Upgrade Upgrade;
  default keep-alive;
}
```

Author: Doesn't work if the request has a multi-value Connection header.

Author: Looks like it might work if I do `"~*Upgrade" Upgrade;`

Reviewer: Given that, I'm fine with the way you have it now or the old way.

Reviewer 2: What does `"~*Upgrade" Upgrade;` mean?

Author: Case-insensitive check for "Upgrade" being in the request header. For example, Firefox sets "keep-alive, Upgrade" which wouldn't match normally.

>   default keep-alive;

Reviewer 2: Comparing to the other nginx doc, why do `default keep-alive;` here and `proxy_http_version 1.1; proxy_set_header   Connection keep-alive;` there?

Author: Not sure what you mean, this is how you set the Connection header to Upgrade for websockets, and keep-alive for other connections. The other doc doesn't handle websockets.

Reviewer 2: Can you not have a generic nginx config that works for normal requests and websockets?

Reviewer 2: If our general nginx guidance doesn't work for websockets we at least need to call that out more clearly there. There's a weird callout there for blazor server apps, but not for general websockets.

>   default keep-alive;

Author: ```suggestion
    default keep-alive;
```
